Add elasticsearch-node tool docs

andrershov opened the pull request:
This PR adds documentation for elasticsearch-node tool https://github.com/elastic/elasticsearch/pull/37696.

Commit:
node-tool docs

elasticmachine commented:
Pinging @elastic/es-distributed

DaveCTurner committed:
Rework docs

andrershov commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -3,81 +3,90 @@
+<<path-settings,`path.data` setting>>. This means that in a disaster you can
+also restart a node by moving its data directories to another host, presuming
+that those data directories can be recovered from the faulty host. Note that it
+is not possible to restore the data directory from a backup because this will
```
@DaveCTurner Not sure what you mean by "is not possible to restore the data directory from a backup". 
If you copy full data folder to another node, this node will be indistinguishable from the previous node.

DaveCTurner replied:
Copying the literal data directory of the dead node somewhere else, post-mortem, is ok, but restoring from a backup (which could be who-knows-how stale) is decidedly not. We occasionally see people taking backups of their data directories and causing hassle when they discover that they can't restore from such things, and I saw a risk that this paragraph might be interpreted wrongly by those kinds of people.

andrershov replied:
@DaveCTurner Now I see what you mean, can we re-phrase it like "Note that if you have previously taken a backup of the data folder of the stopped node, you can not restore from it, you need a data folder state at the moment this node was stopped"

DaveCTurner replied:
The trouble with the phrase "if you have previously taken a backup of the data folder" is that it suggests this is a thing you might try and do, and I think we should avoid making that suggestion. Technically you _can't_ take a backup of the data folder: a backup from which you cannot ever safely restore isn't really a backup at all 🤔.

andrershov replied:
@DaveCTurner I agree. In this case, probably it makes sense to remove this sentence at all because it's confusing?
Also, the better place for explaining that there is no reason to copy data folders because it's not possible to restore from them is on the snapshot page.

Commit:
Add (term, version) instructions and example run

andrershov commented:
@DaveCTurner thanks for you re-work! I've left one question and pushed one commit, that adds (term, version) instructions and updates sample run.
Let me know what you think about the algorithm of running the tool and answering no on each survived master-eligible node. Probably we want to add --output-cluster-state flag instead.

DaveCTurner committed:
Expand section on how to choose the freshest node

DaveCTurner commented:
I was also thinking of a `--dry-run` flag. But in most cases people won't be running this with >1 node remaining (since by far the most common case is 3 master-eligible nodes) so I didn't want to make too big a deal of it. I pushed a change that explains lexicographic ordering using shorter words than "lexicographic" and which pulls this out to a separate paragraph so as not to clutter up the step-by-step instructions. WDYT?

DaveCTurner commented:
@elasticmachine please run elasticsearch-ci/oss-distro-docs

DaveCTurner committed:
Talk about _cluster_ bootstrapping and avoid talking about bootstrapping the node itself

DaveCTurner committed:
Rewording

andrershov reviewed (commented):
@DaveCTurner I still think we need to re-phrase the part about data folder backup, other than that looks good to me.

lcawl commented:
One thing I'd suggest adding is the Synopsis section that lists all the options, per https://www.elastic.co/guide/en/elasticsearch/reference/current/setup-passwords.html
I can take a deeper look when you're ready for doc review.

Commit:
Merge branch 'master' into zen2_node_tool_docs

Commit:
Adjust messages and delimiter to 72 chars

Commit:
Update sample output of unsafe-bootstrap and add detach-cluster

Commit:
Merge branch 'zen2_node_tool_docs' into zen2_node_tool_docs

# Conflicts:
#	docs/reference/commands/node-tool.asciidoc

andrershov commented:
@DaveCTurner I've added `detach-cluster` sample output. Also, I've updated `unsafe-bootstrap` sample output.
One more thing, I have reformatted messages displayed to the user, so they fit 72 chars (same length for the delimiter).
Please note, that I've removed delimiter from the sample output because docs build fails if it's not removed because it confuses this delimiter and delimiter used to show the beginning and end of the source section.

DaveCTurner committed:
Add prose around detach-cluster

DaveCTurner commented:
@andrershov I've added some text around the `detach-cluster` tool and adjusted some other text that was written before we decided that was necessary after unsafe bootstrapping.

The trouble with adjusting the text displayed by the tool is that this is no longer a pure-docs PR and so has to go through a full CI cycle. I suggest splitting those changes out to a separate PR so we can keep the iteration time on this one down and worry less about keeping on top of `master`. My one comment about those changes is that I don't like starting every sentence on a new line - either they should be new paragraphs (separated by a blank line) or else the whole paragraph should be formatted to width. I prefer the latter.

DaveCTurner committed:
Merge branch 'master' into zen2_node_tool_docs

DaveCTurner committed:
Adjust tool output to match #39811

DaveCTurner committed:
More docs work

DaveCTurner committed:
Add synopsis and Parameters sections

DaveCTurner committed:
Separate examples

DaveCTurner commented:
Ok I think this is good to go. I've split it up like the docs for other commands (synopsis/description/parameters/examples) and asked @lcawl for a review. @andrershov would be good to hear from you too.

lcawl committed:
[DOCS] Sorts command reference pages

lcawl commented:
I've sorted the list of commands so it's alphabetical and will take a closer look overall on Monday

andrershov commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -0,0 +1,329 @@
+6. Run the `elasticsearch-node detach-cluster` tool, described below, on every other
+node in the cluster.
+7. Start all other nodes and verify that each one joins the cluster.
+8. Investigate the data in the cluster to discover if any was lost during this
```
@DaveCTurner, do you think it makes sense to add an explicit step "Restore the desired level of fault-tolerance by starting new master-eligible nodes"?

andrershov commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -0,0 +1,329 @@
+To protect against inadvertently joining the wrong cluster, each cluster
+creates a unique identifier, known as the _cluster UUID_, when it first starts
+up. Every node records the UUID of its cluster and will refuse to join a
+cluster with a different UUID in future.
```
Actually, `unsafe-bootstrap` also generates new clusterUUID. However, we only talk about clusterUUID and in `unsafe-bootstrap` section we talk only about resetting voting configuration. I don't hold the opinion strong, but probably we can mention clusterUUID in `unsafe-bootstrap` section.

DaveCTurner replied:
I pushed 8e46c7dcf9416154c3110d56e56dfbe91aaaefd2 to emphasise it's a new cluster formed by `unsafe-bootstrap`, but without defining what this means in terms of the cluster UUID.

andrershov reviewed (commented):
@DaveCTurner looks good. I've left two minor comments.

DaveCTurner committed:
Merge branch 'master' into zen2_node_tool_docs

DaveCTurner committed:
Emphasize thjat it's a _new_ cluster formed by unsafe bootstrapping

DaveCTurner committed:
Merge branch 'zen2_node_tool_docs' of github.com:andrershov/elasticsearch into zen2_node_tool_docs

lcawl commented:
This command reference goes into a lot more detail (particularly in the description) than other such pages. We can add it as-is for now, but in the long term I think it might make sense to move most of that info into a task-related page elsewhere in the Elasticsearch Reference.  For example:

Troubleshooting (or Administering?)
 > Recovering from node failures

What do you think @debadair and @kat257?

DaveCTurner commented:
Thanks @lcawl, I think that would make sense. There's another related tool on the way too which doesn't fit nicely into this page, so we were already contemplating breaking this out elsewhere.

DaveCTurner reviewed (approved):
I'm happy with this for now, thanks to everyone for the reviews.

lcawl commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -0,0 +1,335 @@
+[[node-tool]]
+== elasticsearch-node
+
+The `elasticsearch-node` command allows you to perform unsafe operations that
```
```suggestion
The `elasticsearch-node` command enables you to perform unsafe operations that
```

lcawl commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -0,0 +1,335 @@
+
+Sometimes {es} nodes are temporarily stopped, perhaps because of the need to
+perform some maintenance activity or perhaps because of a hardware failure.
+Once the temporary condition has been resolved you should restart the node and
```
```suggestion
After you resolve the temporary condition and restart the node,
```

lcawl commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -0,0 +1,335 @@
+least one of them is restarted.
+
+In very rare circumstances it may not be possible to restart enough nodes to
+restore the cluster's availability. If such a disaster occurs then you should
```
```suggestion
restore the cluster's availability. If such a disaster occurs, you should
```

lcawl commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -0,0 +1,335 @@
+
+In very rare circumstances it may not be possible to restart enough nodes to
+restore the cluster's availability. If such a disaster occurs then you should
+build a new cluster from a recent snapshot, and re-import any data that was
```
```suggestion
build a new cluster from a recent snapshot and re-import any data that was
```

lcawl commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -0,0 +1,335 @@
+  a brand-new cluster on its own, using its local copy of the cluster metadata.
+  This is known as _unsafe cluster bootstrapping_.
+
+* `elastisearch-node detach-cluster` allows you to move nodes from one cluster
```
```suggestion
* `elastisearch-node detach-cluster` enables you to move nodes from one cluster
```

lcawl commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -0,0 +1,335 @@
+
+* `elastisearch-node detach-cluster` allows you to move nodes from one cluster
+  to another. This can be used to move nodes into the new cluster created with
+  the `elastisearch-node unsafe-bootstap` command above. It also allows you to
```
```suggestion
  the `elastisearch-node unsafe-bootstap` command. It also enables you to
```

lcawl commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -0,0 +1,335 @@
+* `elastisearch-node detach-cluster` allows you to move nodes from one cluster
+  to another. This can be used to move nodes into the new cluster created with
+  the `elastisearch-node unsafe-bootstap` command above. It also allows you to
+  move nodes into a brand-new cluster if unsafe cluster bootstrapping was not
```
```suggestion
  move nodes into a brand-new cluster.
```

lcawl commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -0,0 +1,335 @@
+  to another. This can be used to move nodes into the new cluster created with
+  the `elastisearch-node unsafe-bootstap` command above. It also allows you to
+  move nodes into a brand-new cluster if unsafe cluster bootstrapping was not
+  possible.
```
```suggestion

```

lcawl commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -0,0 +1,335 @@
+to restart a majority of them, then the `elasticsearch-node unsafe-bootstrap`
+command will unsafely override the cluster's <<modules-discovery-voting,voting
+configuration>> as if performing another
+<<modules-discovery-bootstrap-cluster,cluster bootstrapping process>>, allowing
```
```suggestion
<<modules-discovery-bootstrap-cluster,cluster bootstrapping process>>. 
```

lcawl commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -0,0 +1,335 @@
+command will unsafely override the cluster's <<modules-discovery-voting,voting
+configuration>> as if performing another
+<<modules-discovery-bootstrap-cluster,cluster bootstrapping process>>, allowing
+the target node to form a new cluster on its own. This new cluster forms using
```
```suggestion
The target node can then form a new cluster on its own by using
```

lcawl commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -0,0 +1,335 @@
+<<modules-discovery-bootstrap-cluster,cluster bootstrapping process>>, allowing
+the target node to form a new cluster on its own. This new cluster forms using
+the cluster metadata held locally on the target node. This can lead to
+arbitrary data loss since the target node may not hold the latest cluster
```
```suggestion
WARNING: These steps can lead to arbitrary data loss since the target node may not hold the latest cluster
```

lcawl commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -0,0 +1,335 @@
+configuration>> as if performing another
+<<modules-discovery-bootstrap-cluster,cluster bootstrapping process>>, allowing
+the target node to form a new cluster on its own. This new cluster forms using
+the cluster metadata held locally on the target node. This can lead to
```
```suggestion
the cluster metadata held locally on the target node. 

```

lcawl commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -0,0 +1,335 @@
+----
+
+If you have a choice of nodes on which to run this tool then you should choose
+one with a term that is as large as possible, and if there is more than one
```
```suggestion
one with a term that is as large as possible. If there is more than one
```

lcawl commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -0,0 +1,335 @@
+
+If you have a choice of nodes on which to run this tool then you should choose
+one with a term that is as large as possible, and if there is more than one
+node with the same term then you should pick the one with the largest version.
```
```suggestion
node with the same term, pick the one with the largest version.
```

lcawl commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -0,0 +1,335 @@
+If you have a choice of nodes on which to run this tool then you should choose
+one with a term that is as large as possible, and if there is more than one
+node with the same term then you should pick the one with the largest version.
+This identifies the node with the freshest cluster state, minimising the
```
```suggestion
This information identifies the node with the freshest cluster state, which minimizes the
```

lcawl commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -0,0 +1,335 @@
+reports `(4, 10)` then it has the same term as the first node, but has a
+smaller version, so the first node is preferred.
+
+[WARNING]
```
I suggested turning an earlier mention of data loss into a warning, so you can either leave both or combine this with the earlier one.

lcawl commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -0,0 +1,335 @@
+master-eligible nodes are also stopped while this tool is running, but the tool
+does not check this.
+
+[NOTE]
```
```suggestion

```

lcawl commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -0,0 +1,335 @@
+
+To protect against inadvertently joining the wrong cluster, each cluster
+creates a unique identifier, known as the _cluster UUID_, when it first starts
+up. Every node records the UUID of its cluster and will refuse to join a
```
```suggestion
up. Every node records the UUID of its cluster and refuses to join a
```

lcawl commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -0,0 +1,335 @@
+To protect against inadvertently joining the wrong cluster, each cluster
+creates a unique identifier, known as the _cluster UUID_, when it first starts
+up. Every node records the UUID of its cluster and will refuse to join a
+cluster with a different UUID in future.
```
```suggestion
cluster with a different UUID.
```

lcawl commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -0,0 +1,335 @@
+
+However, if a node's cluster has permanently failed then it may be desirable to
+try and move it into a new cluster. The `elasticsearch-node detach-cluster`
+command lets you detach a node from its cluster by resetting its cluster UUID,
```
```suggestion
command lets you detach a node from its cluster by resetting its cluster UUID.
```

lcawl commented on docs/reference/commands/node-tool.asciidoc:
```diff
@@ -0,0 +1,335 @@
+However, if a node's cluster has permanently failed then it may be desirable to
+try and move it into a new cluster. The `elasticsearch-node detach-cluster`
+command lets you detach a node from its cluster by resetting its cluster UUID,
+so that it can join another cluster with a different UUID.
```
```suggestion
It can then join another cluster with a different UUID.
```

lcawl reviewed (approved):
I added minor suggestions but overall LGTM

lcawl committed:
Update docs/reference/commands/node-tool.asciidoc

Co-Authored-By: andrershov <andrershov@gmail.com>

lcawl committed:
Update docs/reference/commands/node-tool.asciidoc

Co-Authored-By: andrershov <andrershov@gmail.com>

lcawl committed:
Update docs/reference/commands/node-tool.asciidoc

Co-Authored-By: andrershov <andrershov@gmail.com>

lcawl committed:
Update docs/reference/commands/node-tool.asciidoc

Co-Authored-By: andrershov <andrershov@gmail.com>

lcawl committed:
Update docs/reference/commands/node-tool.asciidoc

Co-Authored-By: andrershov <andrershov@gmail.com>

lcawl committed:
Update docs/reference/commands/node-tool.asciidoc

Co-Authored-By: andrershov <andrershov@gmail.com>

lcawl committed:
Update docs/reference/commands/node-tool.asciidoc

Co-Authored-By: andrershov <andrershov@gmail.com>

lcawl committed:
Update docs/reference/commands/node-tool.asciidoc

Co-Authored-By: andrershov <andrershov@gmail.com>

lcawl committed:
Update docs/reference/commands/node-tool.asciidoc

Co-Authored-By: andrershov <andrershov@gmail.com>

lcawl committed:
Apply suggestions from code review

Co-Authored-By: andrershov <andrershov@gmail.com>

Commit:
More fixes
